File: src/main/java/com/github/penguin418/todolist/model/entity/TodoEntity.java
package com.github.penguin418.todolist.model.entity;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.github.penguin418.todolist.model.constant.TodoState;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Getter
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
public class TodoEntity extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "todo_id")
    private Long todoId;

    @Setter
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    @JsonManagedReference
    private TodoUserEntity todoUser;

    @Column(name = "todo_content")
    private String todoContent;

    @Column(name = "todo_state")
    private TodoState todoState;

    public void setTodoState(TodoState newState){
        if (!TodoState.isValidTransition(this.todoState, newState)){
            throw new IllegalArgumentException("Invalid state transition");
        }
        this.todoState = newState;
    }
}

File: src/main/java/com/github/penguin418/todolist/model/entity/TodoEntity.kt
package com.github.penguin418.todolist.model.entity

import com.fasterxml.jackson.annotation.JsonManagedReference
import com.github.penguin418.todolist.model.constant.TodoState
import com.github.penguin418.todolist.model.constant.TodoState.Companion.isValidTransition
import jakarta.persistence.*
import lombok.*
import org.springframework.data.jpa.domain.support.AuditingEntityListener

@Entity
@Getter
@AllArgsConstructor
@NoArgsConstructor
@EntityListeners(AuditingEntityListener::class)
class TodoEntity() : BaseEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "todo_id")
     var todoId: Long? = null

    @Setter
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    @JsonManagedReference
     var todoUser: TodoUserEntity? = null

    @Column(name = "todo_content")
     var todoContent: String? = null

    @Column(name = "todo_state")
     var todoState: TodoState? = null
        get() = field
        set(value){
            require(isValidTransition(field, value!!)) { "Invalid state transition" }
            field = value
        }
}
